### PR TITLE
Add tqdm bar

### LINKED ISSUE #4: Show tqdm progress bar for metric functions
I just ran fluency for a list of 100 strings and it was fairly slow. Showing a progress indicator would be nice.

### PR DESCRIPTION
**Pull request by ThanhChinhBK**
close #4 
- [x] added tqdm process bar for some metrics that have for loops inside them

### CODE REVIEW

File: src/langcheck/metrics/en/reference_based_text_quality.py [hunk 1]
```diff
@@ -232,7 +233,7 @@ def _rouge(generated_outputs: List[str], reference_outputs: List[str],
     assert rouge_type in ["rouge1", "rouge2", "rougeLsum"]
     scorer = rouge_scorer.RougeScorer([rouge_type], use_stemmer=True)
     scores = []
-    for gen, ref in zip(generated_outputs, reference_outputs):
+    for gen, ref in tqdm_wrapper(zip(generated_outputs, reference_outputs), total=len(generated_outputs)):
```
Comment (liwii) [inline src/langcheck/metrics/en/reference_based_text_quality.py:236]: I think for consistency we should show the progress bar in all metrics computations. Could you add mini-batching to the other metrics such as local toxicity, and show the progress bar based on the mini batches?

**Review by liwii — COMMENTED** [review]
Thanks for the contribution, I'm very excited to see the progress bar.

Left a few comment.

File: src/langcheck/utils/progess_bar.py [hunk 2]
```diff
@@ -0,0 +1,12 @@
+from tqdm import tqdm
+
+
+def tqdm_wrapper(iterable, desc=None, total=None, unit="it"):
```
[the remark this answers is not in the document]
Comment (ThanhChinhBK) [reply src/langcheck/utils/progess_bar.py:4]: sure, let me add it

File: src/langcheck/metrics/en/reference_based_text_quality.py (hunk 1 above)
Comment (ThanhChinhBK) [reply src/langcheck/metrics/en/reference_based_text_quality.py:236]: sure, it's a nice idea

Comment (ThanhChinhBK) [reply src/langcheck/metrics/en/reference_based_text_quality.py:252]: @liwii I added mini batch, could you take a look

File: src/langcheck/metrics/en/reference_based_text_quality.py [hunk 3]
```diff
@@ -93,19 +93,27 @@ def semantic_similarity(
...
-    # vectors to exceed 1.0 slightly, so we clip the outputs
-    cosine_scores = torch.clamp(cosine_scores, -1.0, 1.0)
+    batch_size = 8
+    scores = []
+    with torch.no_grad():
+        for i in tqdm_wrapper(range(0, len(generated_embeddings), batch_size)):
+            batch_generated_embeddings = generated_embeddings[i:i + batch_size]
+            batch_reference_embeddings = reference_embeddings[i:i + batch_size]
+
+            cosine_scores = util.pairwise_cos_sim(torch.tensor(generated_embeddings),
```
Comment (liwii) [inline src/langcheck/metrics/en/reference_based_text_quality.py:103]: Use `batch_` variables as the arguments?

File: src/langcheck/metrics/en/reference_free_text_quality.py [hunk 4]
```diff
@@ -295,12 +295,16 @@ def _fluency_local(generated_outputs: List[str]) -> List[float]:
...
+    batch_size = 8
+    scores = []
     with torch.no_grad():
-        # Probabilities of [negative, neutral, positive]
-        probs = torch.nn.functional.softmax(
-            _fluency_model(**input_tokens).logits, dim=1)
-
-    return probs[:, 1].tolist()
+        for i in tqdm_wrapper(range(0, len(generated_outputs), batch_size), total=len(generated_outputs)//batch_size):
+            # Probabilities of [negative, neutral, positive]
```
Comment (liwii) [inline src/langcheck/metrics/en/reference_free_text_quality.py:302]: The comment should be placed on line#306

File: src/langcheck/metrics/ja/reference_free_text_quality.py [hunk 5]
```diff
@@ -209,8 +209,13 @@ def _toxicity_local(generated_outputs: List[str]) -> List[float]:
     input_tokens = _toxicity_tokenizer(generated_outputs,
                                        return_tensors='pt',
                                        padding=True)
-    output = _toxicity_model(**input_tokens)
-    toxicity_scores = torch.sigmoid(output.logits[:, 0]).tolist()
+    batchsize = 8
+    toxicity_scores = []
+    for i in tqdm_wrapper(range(0, len(generated_outputs), batchsize), total=len(generated_outputs)//batchsize):
+        with torch.no_grad():
+            output = _toxicity_model(**input_tokens)
```
Comment (liwii) [inline src/langcheck/metrics/ja/reference_free_text_quality.py:216]: Let's use `i` index to make an actual mini-batch here

**Review by liwii — CHANGES_REQUESTED** [review]
Thanks for the further work! Made a few comment by reading the code.

**Comment by liwii** [comment]
Thanks for the fixes!!

I'll take a look again as soon as the CI errors are resolved (we updated the settings so it should run on every new commit automatically).

**Comment by ThanhChinhBK** [comment]
> Thanks for the fixes!!
> 
> I'll take a look again as soon as the CI errors are resolved (we updated the settings so it should run on every new commit automatically).

it seems like the factual_consistency_openai have some bugs, let me check the detail, Sorry, this PR takes to much time

**Comment by liwii** [comment]
No worries, we are thankful for your contribution!!

**Comment by ThanhChinhBK** [comment]
@liwii could you help me on this, I don't understand why yapf failed

File: src/langcheck/metrics/en/reference_based_text_quality.py [hunk 6]
```diff
@@ -71,15 +72,24 @@ def semantic_similarity(
...
         # of all the existing sentence-transformer models that have been
         # evaluated.
         # Ref: https://www.sbert.net/docs/pretrained_models.html#model-overview
         model = SentenceTransformer('sentence-transformers/all-mpnet-base-v2')
-        generated_embeddings = model.encode(generated_outputs)
-        reference_embeddings = model.encode(reference_outputs)
+        generated_embeddings = []
+        reference_embeddings = []
+        for i in tqdm_wrapper(range(0, len(generated_outputs), batch_size),
+                              total=len(generated_outputs) // batch_size,
```
Comment (liwii) [inline src/langcheck/metrics/en/reference_based_text_quality.py:85]: `total` should be `0` when `len(generated_output) == 0`, and `1` when `len(generated_output) == 1`.
```suggestion
                              total=(len(generated_outputs) + batch_size - 1)// batch_size,
```

File: src/langcheck/metrics/en/reference_based_text_quality.py [hunk 7]
```diff
@@ -101,36 +111,60 @@ def semantic_similarity(
...
-    cosine_scores = util.pairwise_cos_sim(torch.tensor(generated_embeddings),
-                                          torch.tensor(reference_embeddings))
-    # Numerical instability can cause the dot product of almost identical
-    # vectors to exceed 1.0 slightly, so we clip the outputs
-    cosine_scores = torch.clamp(cosine_scores, -1.0, 1.0)
+        generated_embeddings = []
+        reference_embeddings = []
+
+        for i in tqdm_wrapper(range(0, len(generated_outputs), batch_size),
+                              total=len(generated_outputs) // batch_size,
```
Comment (liwii) [inline src/langcheck/metrics/en/reference_based_text_quality.py:118]: Same

File: src/langcheck/metrics/en/reference_based_text_quality.py [hunk 8]
```diff
@@ -101,36 +111,60 @@ def semantic_similarity(
...
+            batch_reference_embeddings = [
+                item.embedding for item in batch_ref_embed_response.data
+            ]
+            generated_embeddings.extend(batch_generated_embeddings)
+            reference_embeddings.extend(batch_reference_embeddings)
+
+    scores = []
+    with torch.no_grad():
+        for i in tqdm_wrapper(range(0, len(generated_embeddings), batch_size),
+                              total=len(generated_embeddings) // batch_size,
```
Comment (liwii) [inline src/langcheck/metrics/en/reference_based_text_quality.py:146]: Same

File: src/langcheck/metrics/en/reference_free_text_quality.py [hunk 9]
```diff
@@ -326,12 +328,19 @@ def _fluency_local(generated_outputs: List[str]) -> List[float]:
...
+    batch_size = 8
+    scores = []
     with torch.no_grad():
-        # Probabilities of [negative, neutral, positive]
-        probs = torch.nn.functional.softmax(
-            _fluency_model(**input_tokens).logits, dim=1)
-
-    return probs[:, 1].tolist()
+        for i in tqdm_wrapper(range(0, len(generated_outputs), batch_size),
+                              total=len(generated_outputs) // batch_size):
```
Comment (liwii) [inline src/langcheck/metrics/en/reference_free_text_quality.py:335]: Same

File: src/langcheck/metrics/en/reference_free_text_quality.py [hunk 10]
```diff
@@ -512,7 +522,16 @@ def _toxicity_local(generated_outputs: List[str]) -> List[float]:
     global _toxicity_model
     if _toxicity_model is None:
         _toxicity_model = Detoxify('original')
-    return _toxicity_model.predict(generated_outputs)['toxicity']
+
+    scores = []
+    batch_size = 8
+    for i in tqdm_wrapper(range(0, len(generated_outputs), batch_size),
+                          total=len(generated_outputs) // batch_size):
```
Comment (liwii) [inline src/langcheck/metrics/en/reference_free_text_quality.py:529]: Same

File: src/langcheck/metrics/en/source_based_text_quality.py [hunk 11]
```diff
@@ -196,7 +201,8 @@ def _factual_consistency_local(gen_sentences_list: List[str],
 
     batch_size = 8
     score_list = []
-    for i in range(0, len(model_input_list), batch_size):
+    for i in tqdm_wrapper(range(0, len(model_input_list), batch_size),
+                          total=len(model_input_list) // batch_size):
```
Comment (liwii) [inline src/langcheck/metrics/en/source_based_text_quality.py:205]: Same

File: src/langcheck/metrics/ja/reference_based_text_quality.py [hunk 12]
```diff
@@ -92,23 +93,33 @@ def semantic_similarity(
...
-        reference_embeddings,  # type: ignore[reportGeneralTypeIssues]
-    )
-    # Numerical instability can cause the dot product of almost identical
-    # vectors to exceed 1.0 slightly, so we clip the outputs
-    cosine_scores = torch.clamp(cosine_scores, -1.0, 1.0)
+    batch_size = 8
+    scores = []
+    with torch.no_grad():
+        for i in tqdm_wrapper(range(0, len(generated_outputs), batch_size),
+                              total=len(generated_outputs) // batch_size):
```
Comment (liwii) [inline src/langcheck/metrics/ja/reference_based_text_quality.py:100]: Same

File: src/langcheck/metrics/ja/reference_free_text_quality.py [hunk 13]
```diff
@@ -234,8 +235,17 @@ def _toxicity_local(generated_outputs: List[str]) -> List[float]:
     input_tokens = _toxicity_tokenizer(generated_outputs,
                                        return_tensors='pt',
                                        padding=True)
-    output = _toxicity_model(**input_tokens)
-    toxicity_scores = torch.sigmoid(output.logits[:, 0]).tolist()
+    batchsize = 8
+    toxicity_scores = []
+    for i in tqdm_wrapper(range(0, len(generated_outputs), batchsize),
+                          total=len(generated_outputs) // batchsize):
```
Comment (liwii) [inline src/langcheck/metrics/ja/reference_free_text_quality.py:241]: Same

**Review by liwii — COMMENTED** [review]
Thanks!! It's nice that we have the progress bar everywhere in Japanese & English metrics!!

Just left one minor comment about the minibatch (all the comments are about the same thing), but otherwise LGTM!!

**Review by liwii — APPROVED** [review]
Thanks for the long-run development, LGTM!!
